### DISCUSSION TITLE
global on/where filter criteria with hint

### DISCUSSION THREAD

**Discussion by nico525** [sqlalchemy/sqlalchemy#8305] [Usage Questions]
### Describe the bug

I am currently following the [recipe](https://docs.sqlalchemy.org/en/14/orm/session_events.html#do-orm-execute-global-criteria) from the documentation to add a global filter/where criteria based on a relationship
While doing so, I figured out that table hints do not appear in my database queries. Is there any workaround?

### To Reproduce

```python
@event.listens_for(Session, "do_orm_execute")
def _do_orm_execute(orm_execute_state):
    if (
            orm_execute_state.is_select
            and not orm_execute_state.is_column_load
            and not orm_execute_state.is_relationship_load
    ):

        name="test"
        orm_execute_state.statement = orm_execute_state.statement.options(
            with_loader_criteria(
                HasTimestamp,
                lambda cls: c.children.any(Child.name == name),
                include_aliases=True
            )
        ).with_hint(Child.__table__, "WITH (NOLOCK)")
```

### Error

The hint is not added to the database query

### Versions

- OS: alpine linux
- Python: 3.10
- SQLAlchemy: 1.4.29
- Database: MSSQL
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): pyodbc

**Comment by zzzeek** [comment]
hi -

I can't reproduce your issue as you have only provided an out of context code fragment. I've tried to reconstruct it but I don't know what "c.children" means.    If I just use with_hint(), there's no problem, see below:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import declared_attr
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

Base = declarative_base()


@event.listens_for(Session, "do_orm_execute")
def _do_orm_execute(orm_execute_state):
    if (
        orm_execute_state.is_select
        and not orm_execute_state.is_column_load
        and not orm_execute_state.is_relationship_load
    ):
        name = "test"
        # orm_execute_state.statement = orm_execute_state.statement.options(
        #    with_loader_criteria(
        #        HasTimestamp,
        #        # what's here ???  what's c?
        #        lambda cls: c.children.any(Child.name == name),
        #        include_aliases=True
        #    )
        # ).with_hint(Child.__table__, "WITH (NOLOCK)")
        orm_execute_state.statement = orm_execute_state.statement.with_hint(
            Child.__table__, "WITH (NOLOCK)"
        )


class HasTimestamp:
    @declared_attr
    def children(cls):
        return relationship("Child")


class Parent(HasTimestamp, Base):
    __tablename__ = "parent"

    id = Column(Integer, primary_key=True)
    data = Column(String)


class Child(Base):
    __tablename__ = "child"
    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("parent.id"))
    name = Column(String)


e = create_engine("...", echo=True)
Base.metadata.create_all(e)

s = Session(e)


s.query(Parent).join(Child).all()
```

I get a query like:

```
SELECT parent.id AS parent_id, parent.data AS parent_data 
FROM parent JOIN child WITH (NOLOCK) ON parent.id = child.a_id
```

I would imagine you have a join to Child that is aliased, for which you'd need to set that up explicitly, would need to see your real use case to know how to describe the solution.

**Comment by nico525** [comment]
Sorry, I stripped some code and haven't fully checked it. This is the version which should be runnable. I currently have no access to setup a running script on my computer due to lack of runnable drivers

  
    from sqlalchemy import Column
    from sqlalchemy import create_engine
    from sqlalchemy import event
    from sqlalchemy import ForeignKey
    from sqlalchemy import Integer
    from sqlalchemy import String
    from sqlalchemy.orm import declarative_base, with_loader_criteria
    from sqlalchemy.orm import declared_attr
    from sqlalchemy.orm import relationship
    from sqlalchemy.orm import Session
    
    Base = declarative_base()
    
    
    @event.listens_for(Session, "do_orm_execute")
    def _do_orm_execute(orm_execute_state):
        if (
            orm_execute_state.is_select
            and not orm_execute_state.is_column_load
            and not orm_execute_state.is_relationship_load
        ):
            role = "test"
            orm_execute_state.statement = orm_execute_state.statement.options(
               with_loader_criteria(
                   Parent,
                   lambda c: c.children.any(ProfilesSharedWith.role == role),
                   include_aliases=True
               )
            ).with_hint(ProfilesSharedWith.__table__, "WITH (NOLOCK)")
    
    
    class ProfilesSharedWith(Base):
        __tablename__ = "profiles_shared_with"
        __table_args__ = {"schema": "ui"}
    
        role = Column(String, primary_key=True)
        record_id = Column(ForeignKey("ui.profiles.id"), primary_key=True)
    
    
    class Profile(Base):
        __tablename__ = "profiles"
        __table_args__ = {"schema": "ui"}
    
        id = Column(Integer, primary_key=True)
        name = Column(String)
    
        children = relationship(Child)
    
    e = create_engine("...", echo=True)
    Base.metadata.create_all(e)
    
    s = Session(e)
    r = s.query(Profile).all()


Also, this is the query being generated:

    SELECT ui.profiles.id AS ui_profiles_id, ui.profiles.name AS ui_profiles_name 
    FROM ui.profiles 
    WHERE EXISTS (SELECT 1 
    FROM ui.profiles_shared_with 
    WHERE ui.profiles.id = ui.profiles_shared_with.record_id AND ui.profiles_shared_with.role = ?)


I would expect the NOLOCK being applied like that:

    SELECT ui.profiles.id AS ui_profiles_id, ui.profiles.name AS ui_profiles_name 
    FROM ui.profiles 
    WHERE EXISTS (SELECT 1 
    FROM ui.profiles_shared_with WITH (NOLOCK)
    WHERE ui.profiles.id = ui.profiles_shared_with.record_id AND ui.profiles_shared_with.role = ?)

**Reply by zzzeek** [reply]
you are applying the hint to the outer select() statement, which doesnt include ProfilesSharedWith in the FROM clause. you want this on the inner subquery, so that's where to put with_hint().

now problem, relationship.any() returns an Exists object, that does wrap around a scalar SELECT statement, however, we don't have a with_hint() method available there.   One can be added very simply if you open a feature request, it's a one liner plus a quick test.

But not having that at the moment, we can manually spell out what the any() is doing so that you can get the query you want as follows (note you don't need the `__table__` thing either):

```py
        orm_execute_state.statement = orm_execute_state.statement.options(
            with_loader_criteria(
                Profile,
                #lambda c: c.children.any(ProfilesSharedWith.role == role),
                lambda c: select(ProfilesSharedWith)
                .where(ProfilesSharedWith.record_id == Profile.id)
                .where(ProfilesSharedWith.role == role)
                .with_hint(ProfilesSharedWith, "WITH (NOLOCK)").exists(),
                include_aliases=True,
            )
        )
```

less convenient, but I would assume you only have this query in one place and if you want to pursue getting this method added with a feature request and optional pull request (or we can do it), it's safe for the 1.4 series.

SQL for reference is:

```sql
SELECT profiles.id AS profiles_id, profiles.name AS profiles_name 
FROM profiles 
WHERE EXISTS (SELECT profiles_shared_with.role, profiles_shared_with.record_id 
FROM profiles_shared_with WITH (NOLOCK) 
WHERE profiles_shared_with.record_id = profiles.id AND profiles_shared_with.role = ?)

```

For reference, here's the MCVE if we want to work with this further

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import event, select
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base, with_loader_criteria
from sqlalchemy.orm import declared_attr
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

Base = declarative_base()


@event.listens_for(Session, "do_orm_execute")
def _do_orm_execute(orm_execute_state):
    if (
        orm_execute_state.is_select
        and not orm_execute_state.is_column_load
        and not orm_execute_state.is_relationship_load
    ):
        role = "test"
        orm_execute_state.statement = orm_execute_state.statement.options(
            with_loader_criteria(
                Profile,
                #lambda c: c.children.any(ProfilesSharedWith.role == role),
                lambda c: select(ProfilesSharedWith)
                .where(ProfilesSharedWith.record_id == Profile.id)
                .where(ProfilesSharedWith.role == role)
                .with_hint(ProfilesSharedWith, "WITH (NOLOCK)").exists(),
                include_aliases=True,
            )
        )
        # .with_hint(ProfilesSharedWith.__table__, "WITH (NOLOCK)")


class ProfilesSharedWith(Base):
    __tablename__ = "profiles_shared_with"

    role = Column(String(100), primary_key=True)
    record_id = Column(ForeignKey("profiles.id"), primary_key=True)


class Profile(Base):
    __tablename__ = "profiles"

    id = Column(Integer, primary_key=True)
    name = Column(String)

    children = relationship(ProfilesSharedWith)


e = create_engine(
    "mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/test?driver=ODBC+Driver+13+for+SQL+Server",
    echo=True,
)
Base.metadata.create_all(e)

s = Session(e)
r = s.query(Profile).all()
```

**Reply by nico525** [reply]
explicitly specifying the select statement is indeed working. The only thing I do not like about that solution is I have to maintain the relationship of these objects twice. Once in the relationship property and in the event listener.

if you don't mind adding this as a new feature request that would be great
